keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,799 @@
+
+A webhook server receives AdmissionReview(definition) requests from API server, and responses with a response of the same type to either admit/deny the request. Following simplified diagram shows the workflow.
+(Note that the mutatting webhooks will be invoked BEFORE validating).
+
+The webhook server will expose an HTTP endpoint such that to allow the API server to send AdmissionReview requests. Webhook server providers can dynamically(details) configure what type of resources and what type of admission webhooks via creating CRs of type ValidatingWebhookConfiguration and/or MutatingWebhookConfiguration. It’s worth thinking to provide a generic webhook server in csi-repo. At this stage, the proposal is to create a repo under CSI org, however only implement the validating webhooks for VolumeSnapshot.
+
+CRD validation is preferred over webhook validation due to their lower complexity, however CRD validation schema is unable to enforce immutability or provide ratcheting validation.
+
+### Goals
+
+- Provide an updated CRD schema to validate fields
+- Prevent:
+  - Invalid VolumeSnapshot/VolumeSnapshotContent from creation and update
+  - Invalid updates on immutable fields, i.e., VolumeSnapshot.Spec.Source
+- Provide a pre-built image which can be used to deploy the webhook server
+- Provide a way to deploy the webhook server in cluster
+- Provide a way to authenticate the webhook server to the API server via TLS
+- Provide a release process to safely tighten the validation and move towards the ideal state of using builtin CRD validation while maintaining backwards compatibility
+
+### Non-Goals
+
+- Provide a way to authenticate the API server to the webhook server
+
+## Proposal
+
+Tighten the validation on Volume Snapshot objects. The following fields will begin to enforce immutability: `VolumeSnapshot.Spec.Source` and `VolumeSnapshotContent.Spec.Source`. The following fields will begin to enforce `oneOf`: `VolumeSnapshot.Spec.Source` and `VolumeSnapshotContent.Spec.Source`). The following fields will begin to enforce non-empty strings: `spec.VolumeSnapshotClassName`. More details are in the Validating Scenarios section.
+
+Due to backwards compatibility concerns, the tightening will occur in two phases.
+
+1. The first phase is webhook-only, and will use [ratcheting validation](#backwards-compatibility) combined with a reconciliation method to delete or fix currently persisted objects which are invalid under the new (strict) validation rules.
+2. The second phase occurs once all invalid objects are cleared *from* the cluster. The CRD schema validation will be tightened and the webhook will stick around to enforce immutability until immutable fields come to CRDs. (or the crd upgrade could wait until immutable fields are available to do in one go)
```

Comment:
Do you know when "immutable fields come to CRDs" will happen?

Reply:
https://github.com/kubernetes/enhancements/blob/ae1cfab09d59aafff22f7147c0a5e6b5c6cb38b8/keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md#alternatives

I link to the KEP for the immutable fields for CRDs, as of now it is provisional status. Likely won't be a feature for quite some time.

Reply:
Assuming phase 1 prevents (a) new invalid objects and (b) ensures that once objects are valid, they remain so, is there any benefit to moving to phase 2?
It seems that a benefit only materializes w/ CRD immutable fields, wherein the webhook can be removed entirely. The benefit being: lower complexity and decreased operation latency.

Is the concern that we can't move to v1 until we are sure all invalid objects have been purged?

Reply:
@JohnStrunk the main benefit is not only immutable fields, but also the possibility to add "OneOf" validation into CRD directly instead of relying on validation webhook to do it.
the major issue is that we will NEVER be sure "all invalid objects have been purged" unless we introduce logic into controller to do automatic cleaning up of already existing invalid CRs(which by the way is considered to be dangerous as it might cause data loss). The proposed KEP will rely on users to clean the environment, and we could not guarantee that will happen for sure. The two-phased release steps in this KEP is to mitigate that concern by providing a graceful period to allow users to fix invalid ones.

Reply:
Got it. I think this comes down to me favoring the

> (or the crd upgrade could wait until immutable fields are available to do in one go)

approach.

Reply:
@JohnStrunk I'm not exactly sure what is the requirement for v1. 

I suppose there is a possibility of:

1. Doing phase 1 as described with the ratcheting validation
2. Since the webhook provides the needed validation, it unblocks for GA
3. Then at some later point in time we bring in CRD level validation 
4. Then once CRD immutable fields is stable we can get rid of webhooks entirely because they are (extra complexity, slower)

3 and 4 could happen at the same time, or at different times.

Now this would require changing the CRD schema validation while it is already GA, which may be too much of a change to handle. Although the webhook *should* be enforcing the exact same behaviour as the CRD level schema we will eventually add.

Reply:
GA will happen once phase one and two are both completed, since GA involves the change of the api version anyways, and it is much harder to change the validation after GA.